### PR TITLE
[CI] Integrate E2E tests with GitHub CI

### PR DESCRIPTION
**Pull request by okdas**
## Summary

Allows to automatically run an E2E test on GitHub PR.

### Human Summary

- Changed GitHub workflows to run tests and a build in parallel;
- Added a Kubernetes Job to run the test on the remote cluster;
- Added `E2E_DEBUG_OUTPUT` environment variable to be able to troubleshoot E2E test failures;

### Summary generated by Reviewpad on 10 Nov 23 21:16 UTC
This pull request introduces several changes across multiple files. Here is a summary of the changes:

1. A new file named "run-tests.yml" is added, which includes a new workflow for running tests. The workflow is triggered on push events to the "main" branch and pull requests. Concurrency settings are configured to group workflow runs based on the workflow and branch or pull request. Environment variables "GKE_CLUSTER" and "GKE_ZONE" are set to specific values. The workflow contains a job named "go-test" that runs on the "ubuntu-latest" environment. Several steps are added to the job, including installing the "ignite" tool, checking out the repository with a specified fetch depth, setting up Go with a specified version, installing CI dependencies, generating protobufs, generating mocks, running golangci-lint, and running tests with the "go_test" make command.

2. The Dockerfile.dev has changes related to the installation of `vim` and `less` using `apt-get`. The file path for moving the `poktrolld` binary is changed to `/usr/bin/ignite` and `/usr/bin/poktrolld`. Additionally, port 8546 is exposed in addition to port 8545.

3. The "node.go" file diff adds a new variable called "defaultDebugOutput" and a new condition to set its value based on an environment variable. Additionally, a debug output is printed if the "defaultDebugOutput" variable is true. These changes seem to be related to debugging and adding more control over the output in the code.

4. The `.gitignore` file diff includes changes such as a deleted entry, modified entries, and added entries. Line 32 has a deleted entry: `localnet/*/config/*.json`. Two lines starting with `localnet/poktrolld/config/` have been added: `!localnet/poktrolld/config/client.toml` and `!localnet/poktrolld/config/config.toml`. The entry for ignoring Mac OS files `.DS_Store` remains unchanged.

5. The labels and actions configuration for PRs in a repository is updated. Changes include updating the label and comment when the `devnet-test-e2e` label is added or removed from a PR, updating the label to be removed when `devnet` is removed from a PR, and updating the label to be removed when the `push-image` label is removed.

6. A new file named "run-e2e-test-job-template.yaml" is added, defining a Kubernetes Job that runs end-to-end tests for a Pocket Network application. The Job runs a container using the image "ghcr.io/pokt-network/poktrolld" with specified environment variables. It mounts two volumes for storing configuration files and keys. The Job has a TTL of 120 seconds after it finishes, and the restart policy is set to Never. The backoff limit is set to 0, meaning it will not retry failed attempts.

7. The `reviewdog.yml` file has changes related to concurrency configuration for the workflow, defining the concurrency group based on the workflow and branch name, and enabling the cancellation of in-progress jobs. There is also a comment about blocking certain comments to ensure the merging to the `main` branch is not affected.

8. A new shell script named "run-e2e-test.sh" is added, which sets up a sequence of checks and actions to run end-to-end tests in a Kubernetes environment. It waits for a pod with the matching image SHA and purpose to become ready, creates a job to run the e2e tests, monitors the logs and status of the pod, and deletes the job based on the pod's status and exit code.

9. The file diff involves renaming the file from "go.yml" to "main-build.yml", updating the workflow name from "Ignite build & test" to "Main build", modifying the concurrency group to include the workflow name, changing the job name for building the container to "build-push-container", updating the Go version to "1.20.10", removing steps for generating mocks and running golangci-lint, changing the Docker image name from "pocketd" to "poktrolld", modifying the step for copying binaries to include the ignite binary and the poktrolld binary, and adding a new job named "run-e2e-tests" that runs E2E tests based on the "devnet-test-e2e" label.

Overall, this pull request introduces a range of changes related to running tests, Dockerfile modifications, debugging, gitignore configuration, labels and actions for PRs, Kubernetes job templates, reviewdog configuration, shell script for e2e testing, and workflow updates.

## Issue

- #18

## Type of change

Select one or more:

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

- [x] **Run all unit tests**: `make go_develop_and_test`
- [x] **Verify Localnet manually**: See the instructions [here](TODO: add link to instructions)

## Sanity Checklist

- [ ] I have tested my changes using the available tooling
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, updated documentation and left TODOs throughout the codebase

### CODE REVIEW

File: .github/workflows-helpers/run-e2e-test.sh [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+PODS_JSON=$(kubectl get pods -n ${NAMESPACE} -l pokt.network/purpose=sequencer -o json)
+
+# Check if any pods are running and have the correct image SHA
+READY_POD=$(echo $PODS_JSON | jq -r ".items[] | select(.status.phase == \"Running\") | select(.spec.containers[].image | contains(\"${IMAGE_TAG}\")) | .metadata.name")
+
+if [[ -n "${READY_POD}" ]]; then
+    echo "Ready pod found: ${READY_POD}"
+    break
+else
+    echo "Sequencer with with an image ${IMAGE_TAG} is not ready yet. Waiting..."
```
Comment (Olshansk) [inline .github/workflows-helpers/run-e2e-test.sh:14]:
```suggestion
    echo "Sequencer with with an image ${IMAGE_TAG} is not ready yet. Will retry in 10 seconds..."
```

File: .github/workflows-helpers/run-e2e-test.sh [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    sleep 10
+fi
+done
+
+# Create a job to run the e2e tests
```
Comment (Olshansk) [inline .github/workflows-helpers/run-e2e-test.sh:19]: Great comments throughout!

File: e2e/tests/node.go [hunk 3]
```diff
@@ -30,6 +32,12 @@ func init() {
 	if defaultHome == "" {
 		defaultHome = "../../localnet/poktrolld"
 	}
+
+	if os.Getenv("E2E_DEBUG_OUTPUT") == "true" {
```
Comment (Olshansk) [inline e2e/tests/node.go:36]: You should retrieve the env variable on line 25, similar to `defaultHome`

File: .github/workflows-helpers/run-e2e-test-job-template.yaml [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+          value: celestia-rollkit
+        volumeMounts:
+        - mountPath: /root/.pocket/keyring-test/
+          name: keys-volume
+        - mountPath: /root/.pocket/config/
+          name: configs-volume
+      restartPolicy: Never
+      volumes:
+      - configMap:
+          defaultMode: 420
```
Comment (Olshansk) [inline .github/workflows-helpers/run-e2e-test-job-template.yaml:37]: towelie-south-park

File: .github/workflows-helpers/run-e2e-test.sh [hunk 5]
```diff
@@ -0,0 +1,55 @@
+# Check if the pod with the matching image SHA and purpose is ready
```
Comment (Olshansk) [inline .github/workflows-helpers/run-e2e-test.sh:1]: Not a blocker but just a question.

Is the plan to keep this bash script as is? I know you've used things like argoCD and other DAG like workflow builders, so was just surprised to have a simple bash script like it today.

File: .github/workflows/run-tests.yml [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+          ignite version
+
+      - uses: actions/checkout@v3
+        with:
+          fetch-depth: "0" # Per https://github.com/ignite/cli/issues/1674#issuecomment-1144619147
+
+      - name: Set up Go
+        uses: actions/setup-go@v4
+        with:
+          go-version: "1.20"
```
Comment (Olshansk) [inline .github/workflows/run-tests.yml:33]: Should we be consistent?

```suggestion
          go-version: "1.20.10"
```

**Review by Olshansk — APPROVED** [review]
No blockers, just some comments. I don't really know how to review something like this given that it works 😅

Great job! I know how difficult all these configs, tests, permissions, etc are!!!

File: .github/workflows-helpers/run-e2e-test.sh (hunk 2 above)
Comment (okdas) [reply .github/workflows-helpers/run-e2e-test.sh:19]: thanks @openai

**Pull request merged**